### ISSUE TITLE
Work log

### ISSUE THREAD

**Issue by beatobongco** [beatobongco/sentinel#1]
Todo next time

* Remember sys must try not be bother operators so bigger chance it be used. It's mainly a deterrent. Can argue for migs that they already have pics on file
* ~Get euclidean distance to work on KNOWN and UNKNOWN and observe accuracy~ done
* ~Find ways to store userdata efficiently (is it an image, a tensor?)~ done -- just the embeddings

--

TIL
* remember what we're trying to accomplish, it must be a system that works. 
* I want this to be used, so perhaps can deploy for free in CCM
* could be simply a laptop with webcam mounted up in front where users must show their face for a second, when system registers face a sound plays, face is added to db (have to find easy persistent way, maybe firebase), then we just use euclidean distance each time. If not recognized, sound plays and person must input full name
* for even better performance on systems with poor hardware, the system could only get the first good face, then do classification
* people in CCM scan their tags anyway
* perhaps run on a flask server that I host myself with simple security but ofc focus on client side proof of concept first. servers last. 
* skateboard mentality. for POC, one face at a time is OK and camera (not video) mode. You can freeze the feed too so it looks like ur taking a pic

PROS
* runs good enough on cheap hardware
* library is flexible enough to allow granulated control, features can be turned on/off for hardware constraints

CONS
* This system isn't great at detecting faces at a weird angle
* I cant expect my users to have good hardware

* you can get the tensors via `faceapi.extractFaceTensors(input, detections)`, could be useful if we need to run a SVC

**Comment by beatobongco** [comment]
Next:
Euclidean distance doesn't perform too well. I'll try making a pipeline to train a SVC on face embeddings.

Another option is a simple neural network in tfjs

* https://github.com/karpathy/svmjs
* https://github.com/mljs/libsvm

Also I should try the more powerful face detection models later on (and make them available in single shot mode or just in general), but first get MTCNN defaults to work at least a bit.

**Comment by beatobongco** [comment]
Or create the ability for people to create training data from the app itself, take like 100 shots, store in localstorage then run euclidean distance.

Might not be good for when num images increases + realtime but should work for singleshot.

Real time could work by deferring classification, just save all images of people walk in and if euclidean distance is near for subsequent, dont store. Kind of like we only want unique pics in our db.

**Comment by beatobongco** [comment]
Got training and inferencing to work today!! Data is also saved and retrieved in localstorage.

**Comment by beatobongco** [comment]
Today 
* refactored to use a db object https://github.com/beatobongco/TIL/blob/master/day_notes/2017-10-23_OLOO_Pattern.js
* learned to use async/await correctly

**Comment by beatobongco** [comment]
Today
* Now keeps embeddings of faces it hasn't seen and tags them with an autoincrementing `Unknown #`
* Ported some of the UI to Vue
* Initial prototype of class editor

Next steps
* make it so that when db state changes, props are passed to children so everything updates right.
  * make db pass classes down to main app 
  * think about converting db into a vue obj or refactoring so that db is the actual app
  * https://vuejs.org/v2/guide/state-management.html#Simple-State-Management-from-Scratch
* Fix slider value display

Takeaways
* `className` seems to be a reserved word
* Vue is really close to OLOO pattern or at least easy to make Vue objs from OLOOs

**Comment by beatobongco** [comment]
TODO

* create sleep time for face detection to prevent multiple for each class being taken
* create merging function that merges close faces together
* ~create warmup function to get network warmed up, do forward pass on bogus image or something... `Warming up the system...`~

**Comment by beatobongco** [comment]
* Vincent has an even faster model avail. Prioritize other features first tho, the 3 above. https://github.com/justadudewhohacks/face-api.js/tree/tiny-yolov2-seperable-conv2d

**Comment by beatobongco** [comment]
* Autotrain: Make training take single snaps to maximize variance and minimize num training needed, too much button pressing is fiddly.
* To max variance just make sure distance is greater than a certain threshold
* I have to do quite a shitton of refactoring. Remove boolean flags, make everything a Vue app and make things stateful, dont jQuery style this.
* To start with, create a `training` component
* Hide controls and links while loading

**Comment by beatobongco** [comment]
I have to do quite a shitton of refactoring

**Comment by beatobongco** [comment]
The shit lags too much. Best hide the controls if I cant even use them.

**Comment by beatobongco** [comment]
I need a blur filter
